e2e/single-cluster/finalizers_test.go

```diff
@@ -0,0 +1,296 @@
+		)
+	})
+
+	When("deleting an existing GitRepo", func() {
+		JustBeforeEach(func() {
+			By("creating a GitRepo")
+			err := testenv.CreateGitRepo(k, targetNamespace, gitrepoName, "master", path)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("updates the deployment", func() {
+			By("checking the bundle and bundle deployment exist")
+			Eventually(func() string {
+				out, _ := k.Get("bundles")
+				return out
+			}).Should(ContainSubstring(gitrepoName))
+
+			Eventually(func() string {
+				out, _ := k.Get("bundledeployments", "-A")
+				return out
+			}).Should(ContainSubstring(gitrepoName))
+
+			By("scaling down the Fleet controller to 0 replicas")
+			_, err := k.Namespace("cattle-fleet-system").Run(
+				"scale",
+				"deployment",
+				"fleet-controller",
+				"--replicas=0",
+				"--timeout=5s",
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("deleting the GitRepo")
+			out, err := k.Delete("gitrepo", gitrepoName, "--timeout=2s")
+			Expect(err).To(HaveOccurred())
+			Expect(out).To(ContainSubstring("timed out"))
+
+			By("checking that the gitrepo still exists and has a deletion timestamp")
+			out, err = k.Get(
+				"gitrepo",
+				gitrepoName,
+				"-o=jsonpath={range .items[*]}{.metadata.deletionTimestamp}",
+			)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(out).ToNot(BeZero())
+
+			By("checking that the bundle and bundle deployment still exist")
+			out, err = k.Get("bundles")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(out).To(ContainSubstring(gitrepoName))
+
+			out, err = k.Get("bundledeployments", "-A")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(out).To(ContainSubstring(gitrepoName))
+
+			By("checking that the auxiliary resources still exist")
+			serviceAccountName := fmt.Sprintf("git-%s", gitrepoName)
+			Consistently(func() error {
+				out, _ := k.Get("configmaps")
+				if !strings.Contains(out, fmt.Sprintf("%s-config", gitrepoName)) {
+					return errors.New("configmap not found")
+				}
+
+				out, _ = k.Get("serviceaccounts")
+				if !strings.Contains(out, serviceAccountName) {
+					return errors.New("serviceaccount not found")
+				}
+
+				out, _ = k.Get("roles")
+				if !strings.Contains(out, serviceAccountName) {
+					return errors.New("role not found")
+				}
+
+				out, _ = k.Get("rolebindings")
+				if !strings.Contains(out, serviceAccountName) {
+					return errors.New("rolebinding not found")
+				}
+
+				return nil
+			}, 2*time.Second, 100*time.Millisecond).ShouldNot(HaveOccurred())
+
+			By("deleting the GitRepo once the controller runs again")
+			_, err = k.Namespace("cattle-fleet-system").Run(
+				"scale",
+				"deployment",
+				"fleet-controller",
+				"--replicas=1",
+				"--timeout=5s",
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = k.Delete("gitrepo", gitrepoName)
+			Expect(err).NotTo(HaveOccurred())
+
+			// These resources should be deleted when the GitRepo is deleted.
+			By("checking that the auxiliary resources don't exist anymore")
```

Comment:
We could also check that the `bundle`s and `bundledeployment`s are deleted

Reply:
This does not depend on finalizers though, but rather on relationships between a `GitRepo`, bundles and bundle deployments. We cover some of it [here](https://github.com/rancher/fleet/blob/main/e2e/single-cluster/gitrepo_test.go#L78).